lib/checkother.cpp

```diff
@@ -1445,7 +1462,7 @@ void CheckOther::checkConstVariable()
                 continue;
         }
         // Skip if address is taken
-        if (Token::findmatch(var->nameToken(), "& %varid%", scope->bodyEnd, var->declarationId()))
+        if (hasBadAddressOf(var, scope))
```

Comment:
Just checking the returns is not sufficient, see test cases here: https://github.com/danmar/cppcheck/blob/19eef2c5844d45e4ccc361939e1814c91bc58e5d/test/testother.cpp#L2583

Reply:
So I linked to the wrong line. The problem is with the test just below, which produces a FP:
~~~
        check("struct T : public U { void dostuff() const {}};\n"
              "void a(T& x) {\n"
              "    x.dostuff();\n"
              "    U * const y = dynamic_cast<U * const>(&x)\n"
              "    y->mutate();\n" //to avoid warnings that y can be const
              "}");
~~~
I think the `castToNonConst` check is looking at the wrong bit in `constness`, and also doesn't take the type of `var` into account.